Add explainer for navigation with Web Bundles (#512)
WICG/webpackage: explainers/navigation-to-bundles.md

# Explainer: Navigation with Web Bundles<br>(a.k.a. Bundled HTTP Exchanges)

Last updated: Oct 29, 2019

> (This explainer was originally written and published as: https://docs.google.com/document/d/1KFmtiE3DHgKfQH5-nKtLiacMrXsoKIXQZ-VIMGHMje0/edit?pli=1#)

Reviewer: I think we can leave this back-reference in the commit message instead of the body.

Author: Good idea, done.

A [demo video](https://www.youtube.com/watch?v=rs-3R0ji6dA) that tries to show one of the use-case scenarios of Web Bundles (a.k.a. Bundled HTTP Exchanges):

<a href="https://www.youtube.com/watch?v=rs-3R0ji6dA">
<img src="https://img.youtube.com/vi/rs-3R0ji6dA/0.jpg">
</a>

> Today, loading a website basically means retrieving multiple resources from one or multiple web servers. This gives web a great strength of linkable, indexable, composable and ephemeral, but it is also making it difficult for a website to:

Reviewer: ```suggestion
Today, loading a website basically means retrieving multiple resources from one or multiple web servers. This gives the web the great strengths of being linkable, indexable, composable and ephemeral, but it also makes it difficult for a website to:
```

* be shared and loaded in a self-contained way, similar to what packaged apps and other portable data representation formats provide,
* be composed as a distributable, installable application, or
* load reliably and quickly even when a site consists of a large number of subresources, and even when the internet access is limited.

> Instead, imagine if we could bundle up a “full” website in a single resource file, so that the website could be shared via a SD card or over some p2p protocol, or could be retrieved from a fast cache or a nearby proxy. It would open up various interesting use cases.

Reviewer: I think you can drop the quotes around "full". This phrasing doesn't imply that you *have to* include the full website, just that it's possible.

```suggestion
Instead, imagine if we could bundle up a full website in a single resource file, so that the website could be shared via a SD card or over some p2p protocol, or could be retrieved from a fast cache or a nearby proxy. It would open up several interesting use cases.
```

> **Web Bundle** is a proposal that tries to achieve it. It’s a format that can represent a collection of HTTP resources, and therefore can represent one or multiple web pages (which typically consist of multiple resources, e.g. HTML, javascripts, images and styles) in a single file.  It’s a part of [Web Packaging](https://github.com/WICG/webpackage) proposal and is also known as "**Bundled HTTP Exchanges** ([spec proposal](https://wicg.github.io/webpackage/draft-yasskin-wpack-bundled-exchanges.html))."

Reviewer: ```suggestion
The **Web Bundles** proposal tries to achieve this. It’s a format that can represent a collection of HTTP resources, and therefore can represent one or multiple web pages in a single file, including their subresources like scripts, images and styles.  It’s a part of the [Web Packaging](https://github.com/WICG/webpackage) project and is also known as "**Bundled HTTP Exchanges**" ([spec proposal](https://wicg.github.io/webpackage/draft-yasskin-wpack-bundled-exchanges.html)).
```

> This document explains how a Web Bundle can be a navigation target, so that a user can open a given Web Bundle in browsers (UAs) and navigate into the website represented by the Bundle.

Reviewer: ```suggestion
This document explains how a browser can navigate to a Web Bundle so that a user can open one.
```

> ## Example Scenarios: Save and browse an unsigned content

Reviewer: ```suggestion
## Example Scenarios: Save and browse unsigned content
```

> UA can provide **‘Save as bundle’** feature that will dynamically generate an unsigned bundle of the page that a user is currently browsing.  Then anyone can use the feature to get a bundled representation of the current page, which can be browsed later by the same user, or can be shared with a nearby friend via some person-to-person sharing mechanism.

Reviewer: ```suggestion
A UA can provide a **'Save as bundle'** feature that dynamically generates an unsigned bundle of the page that its user is currently browsing.  Then anyone can create a bundled representation of the current page, and they can browse it later themself or share it with a nearby friend via one of several native file-sharing apps.
```

> When someone else browses the unsigned bundle, they see some notation in their URL bar that isn’t the URL of the original URL, but they can still browse around and see the JS executing.  Note that in this case the page coming from the unsigned bundle is not given any access to the cookie/storage of the original site.

Reviewer: @estark37 commented that showing the original URL is probably a bad idea. It sounds like we'll show the package's URL and/or a chip describing what's happening.

```suggestion
When someone else browses the unsigned bundle, they can browse around and see the site, including the results of its JavaScript executing.  Note that in this case the page inside the unsigned bundle is not given any access to the cookies or storage of the original site.
```

> ## Example Scenarios: Publishing a signed content

Reviewer: ```suggestion
## Example Scenarios: Publishing signed content
```

> The author of a site creates a bundle for one or a few pages of the site, signs the bundle with the site’s certificate, and then publish the bundle in the way that interested users can find it.

Reviewer: ```suggestion
The author of a site creates a bundle for part or all of their site, signs the bundle with the site’s certificate, and then publishes the bundle in a way that interested users can find it. Users can then share this bundle peer-to-peer until its expiration time.
```

> Later, when a user finds the bundle, they can navigate their UA client to the bundle, maybe by opening it as a local file copied over SD card or by navigating to the distribution URL where the bundle is published.  The UA should be able to parse and verify the bundle’s signature, and then to navigate to the website represented by the bundle, without actually connecting to the site as all the necessary subresources could be served by the bundle.  If the bundle represents multiple pages for the site, the user should also be able to browser those pages without worrying about the connectivity.

Reviewer: ```suggestion
Later, when a user receives the bundle, maybe by opening it from an SD card or by navigating to the distribution URL where the bundle is published, they can open it with their UA.  The UA parses and verifies the bundle’s signature, and then opens the pages inside the bundle, without actually connecting to the site's server, but with assurance that it's safe to show the real site in the URL bar and use the site's local storage.  If the bundle represents multiple pages for the site, the user can browse those pages without worrying about connectivity.
```

> # Navigation

Reviewer: ```suggestion
# Details of how the browser handles navigation
```

I'm not doing a detailed review of the text below, because it's intrinsically more technical than we want to make an explainer's audience read. It'll eventually move into the specification.

If a request is a navigation request or a worker request and if the response was from the BundledResources, the initiator’s bundledResources are copied to the target browsingContext.

## Unique-origin Web Bundle Loading
When a Bundle without a valid signature is loaded from a unique origin (e.g. a local file), following URL conversion must be performed:

* Set the document’s URL to the unique URL of the Bundle package's one. 
   * See [@jyasskin](https://github.com/jyasskin)'s [Origins for Resources in Unsigned Packages](https://docs.google.com/document/d/1BYQEi8xkXDAg9lxm3PaoMzEutuQAZi1r8Y0pLaFJQoo/edit#heading=h.1fej4450b9k9) to see what's being proposed.
   * Chrome Canary 80 has an [experimental implementation of this feature](https://chromium.googlesource.com/chromium/src/+/refs/heads/master/content/browser/web_package/using_web_bundles.md), and it creates a URL for a local Bundle by concatenation of the location (URL) of the Bundle, `?` and the primaryUrl of the Bundle (that must be properly URL-encoded). (e.g. if the primaryUrl is `https://bar.com/article.html`, the document’s URL would be shown as `file:///foo/bar.wbn?https://bar.com/article.html`.)

>    * TODO: also look at what TAG did in the original packaging spec: https://www.w3.org/TR/2015/WD-web-packaging-20150115/#fragment-identifiers

Reviewer: This is incorporated into https://docs.google.com/document/d/1BYQEi8xkXDAg9lxm3PaoMzEutuQAZi1r8Y0pLaFJQoo/edit, so I think we don't need it here. 

```suggestion
```
